uaiowebsockets/protocol.py

```diff
@@ -0,0 +1,248 @@
+class Websocket:
+    """
+    Basis of the Websocket protocol.
+
+    This can probably be replaced with the C-based websocket module, but
+    this one currently supports more options.
+    """
+    is_client = False
+
+    def __init__(self, sock, sreader, swriter):
+        self.sock = sock
+        self.sreader = sreader
+        self.swriter = swriter
+        self.open = True
+
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, exc_type, exc, tb):
+        await self.close()
+
+    def settimeout(self, timeout):
+        self.sock.settimeout(timeout)
+
+    async def read_frame(self, max_size=None):
+        """
+        Read a frame from the socket.
+        See https://tools.ietf.org/html/rfc6455#section-5.2 for the details.
+        """
+
+        # Frame header
+        two_bytes = await self.sreader.read(2)
+
+        if not two_bytes:
+            raise NoDataException
+
+        byte1, byte2 = struct.unpack('!BB', two_bytes)
+
+        # Byte 1: FIN(1) _(1) _(1) _(1) OPCODE(4)
+        fin = bool(byte1 & 0x80)
+        opcode = byte1 & 0x0f
+
+        # Byte 2: MASK(1) LENGTH(7)
+        mask = bool(byte2 & (1 << 7))
+        length = byte2 & 0x7f
+
+        if length == 126:  # Magic number, length header is 2 bytes
+            length, = struct.unpack('!H', self.sock.read(2))
+        elif length == 127:  # Magic number, length header is 8 bytes
+            length, = struct.unpack('!Q', self.sock.read(8))
+
+        if mask:  # Mask is 4 bytes
+            mask_bits = await self.sreader.read(4)
+
+        try:
+            data = await self.sreader.read(length)
+        except MemoryError:
+            # We can't receive this many bytes, close the socket
+            if __debug__: LOGGER.debug("Frame of length %s too big. Closing",
+                                       length)
+            self.close(code=CLOSE_TOO_BIG)
+            return True, OP_CLOSE, None
+
+        if mask:
+            data = bytes(b ^ mask_bits[i % 4]
+                         for i, b in enumerate(data))
+
+        return fin, opcode, data
+
+    async def write_frame(self, opcode, data=b''):
+        """
+        Write a frame to the socket.
+        See https://tools.ietf.org/html/rfc6455#section-5.2 for the details.
+        """
+        fin = True
+        mask = self.is_client  # messages sent by client are masked
+
+        length = len(data)
+
+        # Frame header
+        # Byte 1: FIN(1) _(1) _(1) _(1) OPCODE(4)
+        byte1 = 0x80 if fin else 0
+        byte1 |= opcode
+
+        # Byte 2: MASK(1) LENGTH(7)
+        byte2 = 0x80 if mask else 0
+
+        if length < 126:  # 126 is magic value to use 2-byte length header
+            byte2 |= length
+            await self.swriter.awrite(struct.pack('!BB', byte1, byte2))
+
+        elif length < (1 << 16):  # Length fits in 2-bytes
+            byte2 |= 126  # Magic code
+            await self.srwriter.awrite(struct.pack('!BBH', byte1, byte2, length))
```

Comment:
Not sure of the state of this PR but just noticed this typo when I was playing around with it on my Raspberry Pi Pico W.
```suggestion
            await self.swriter.awrite(struct.pack('!BBH', byte1, byte2, length))
```